share/shwap/p2p/shrex/server.go

```diff
@@ -0,0 +1,216 @@
+		host:       host,
+		params:     params,
+		middleware: middleware,
+	}
+
+	protocols := make([]protocol.ID, len(registry))
+	index := 0
+	for protocolName, id := range registry {
+		handler := srv.streamHandler(srv.ctx, id)
+		withRateLimit := srv.middleware.rateLimitHandler(ctx, handler, id().Name())
+		withRecovery := RecoveryMiddleware(withRateLimit)
+
+		protocols[index] = ProtocolID(srv.params.NetworkID(), protocolName)
+		srv.SetHandler(protocols[index], withRecovery)
+		index++
+	}
+	return srv, nil
+}
+
+func (srv *Server) SetHandler(p protocol.ID, h network.StreamHandler) {
+	srv.host.SetStreamHandler(p, h)
+}
+
+// Stop stops the Server
+func (srv *Server) Stop(_ context.Context) error {
+	srv.cancel()
+	for _, protocol := range srv.protocols {
+		srv.host.RemoveStreamHandler(protocol)
+	}
+	return nil
+}
+
+func (srv *Server) Protocols() []protocol.ID {
+	return srv.protocols
+}
+
+func (srv *Server) WithMetrics() error {
+	metrics, err := InitServerMetrics()
+	if err != nil {
+		return fmt.Errorf("shrex/server: init Metrics: %w", err)
+	}
+	srv.metrics = metrics
+	return nil
+}
+
+func (srv *Server) streamHandler(ctx context.Context, id newRequestID) network.StreamHandler {
+	return func(s network.Stream) {
+		requestID := id()
+		handleTime := time.Now()
+		status := srv.handleDataRequest(ctx, requestID, s)
+		srv.metrics.observeRequests(ctx, 1, requestID.Name(), status, time.Since(handleTime))
+		log.Debugw("server: handling request",
+			"name", requestID.Name(),
+			"status", status,
+			"duration", time.Since(handleTime),
+		)
+		// reset because we will not send anything back
+		if status == statusBadRequest || status == statusReadReqErr {
+			s.Reset() //nolint:errcheck
+			return
+		}
+
+		if err := s.Close(); err != nil {
+			log.Debugw("server: closing stream", "err", err)
+		}
+	}
+}
+
+func (srv *Server) handleDataRequest(ctx context.Context, requestID request, stream network.Stream) status {
+	logger := log.With(
+		"source", "server",
+		"name", requestID.Name(),
+		"peer", stream.Conn().RemotePeer().String(),
+	)
+
+	logger.Debugf("handling data request: %s", requestID.Name())
+
+	err := stream.SetReadDeadline(time.Now().Add(srv.params.ReadTimeout))
+	if err != nil {
+		logger.Debugw("setting read deadline", "err", err)
+	}
+
+	_, err = requestID.ReadFrom(stream)
+	if err != nil {
+		logger.Warnf("reading request: %w", err)
+		return statusReadReqErr
+	}
+
+	err = stream.CloseRead()
+	if err != nil {
+		logger.Warnw("closing read side of the stream", "err", err)
+	}
+
+	err = requestID.Validate()
+	if err != nil {
+		logger.Warnw("validate request", "err", err)
+		return statusBadRequest
+	}
+
+	logger.Debugw("new request")
+	ctx, cancel := context.WithTimeout(ctx, srv.params.HandleRequestTimeout)
+	defer cancel()
+
+	var r io.Reader
+	file, err := srv.store.GetByHeight(ctx, requestID.Height())
+	if err == nil {
+		defer utils.CloseAndLog(log, "file", file)
+		r, err = requestID.ResponseReader(ctx, file)
+	}
```

Comment:
Should this be after switch case (line 177)? Will be much easier to read if return fast statements moved above.

Reply:
will rework it

Reply:
Consider handle them separately? combining this handling makes code very unusual and hard to read